### ISSUE TITLE
ArrayIndexOutOfBoundsException in KEDetectedFace

### ISSUE THREAD

**Issue by jonhare** [openimaj/openimaj#67]
OP-68: Hello OpenIMAJ Team!

I hope you are doing well, and since the Santa Claus is coming I wish all of you a wonderful Christmas time spent with your families :)

I'm not sure if I selected correct component for this issue, but here is the problem. 

I'm getting array out of bounds exception:

{code}
Exception in thread "AWT-EventQueue-0" java.lang.ArrayIndexOutOfBoundsException: 9
	at org.openimaj.image.processing.face.detection.keypoints.KEDetectedFace.getKeypoint(KEDetectedFace.java:88)
	at com.github.sarxos.webcam.FacePainterExample.paintImage(FacePainterExample.java:116)
	at com.github.sarxos.webcam.WebcamPanel.paintComponent(WebcamPanel.java:543)
	at javax.swing.JComponent.paint(JComponent.java:1037)
	at javax.swing.JComponent.paintToOffscreen(JComponent.java:5132)
	at javax.swing.BufferStrategyPaintManager.paint(BufferStrategyPaintManager.java:278)
	at javax.swing.RepaintManager.paint(RepaintManager.java:1257)
	at javax.swing.JComponent._paintImmediately(JComponent.java:5080)
	at javax.swing.JComponent.paintImmediately(JComponent.java:4890)
	at javax.swing.RepaintManager$3.run(RepaintManager.java:814)
	at javax.swing.RepaintManager$3.run(RepaintManager.java:802)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.AccessControlContext$1.doIntersectionPrivilege(AccessControlContext.java:87)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:802)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:745)
	at javax.swing.RepaintManager.prePaintDirtyRegions(RepaintManager.java:725)
	at javax.swing.RepaintManager.access$1000(RepaintManager.java:46)
	at javax.swing.RepaintManager$ProcessingRunnable.run(RepaintManager.java:1668)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:209)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:672)
	at java.awt.EventQueue.access$400(EventQueue.java:81)
	at java.awt.EventQueue$2.run(EventQueue.java:633)
	at java.awt.EventQueue$2.run(EventQueue.java:631)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.AccessControlContext$1.doIntersectionPrivilege(AccessControlContext.java:87)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:642)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:269)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:184)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:174)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:169)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:161)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:122)
{code}

My code:

{code}
List<KEDetectedFace> faces = DETECTOR.detectFaces(ImageUtilities.createFImage(webcam.getImage()));

if (faces == null) {
	return;
}

Iterator<KEDetectedFace> dfi = faces.iterator();
while (dfi.hasNext()) {

	KEDetectedFace face = dfi.next();
	Rectangle bounds = face.getBounds();

	int x = (int) bounds.x;
	int y = (int) bounds.y;
	int w = (int) bounds.width;
	int h = (int) bounds.height;

	FImage path = face.getFacePatch();

	FacialKeypoint pel = face.getKeypoint(FacialKeypointType.EYE_LEFT_CENTER);
	FacialKeypoint per = face.getKeypoint(FacialKeypointType.EYE_RIGHT_CENTER);
	FacialKeypoint pmc = face.getKeypoint(FacialKeypointType.MOUTH_CENTER);

	g2.drawImage(ImageUtilities.createBufferedImage(path), 0, 0, null);
	g2.setStroke(STROKE);
	g2.setColor(Color.RED);
	g2.drawRect(x, y, w, h);
	g2.setColor(Color.GREEN);

	if (pel != null) {
		g2.drawRect((int) pel.position.x, (int) pel.position.y, 1, 1);
	}
	if (per != null) {
		g2.drawRect((int) per.position.x, (int) per.position.y, 1, 1);
	}
}
{code}

The problem is on this line:

{code}
FacialKeypoint pel = face.getKeypoint(FacialKeypointType.EYE_LEFT_CENTER);
{code}

And from the OpenIMAJ source code:

{code}
public FacialKeypoint getKeypoint(FacialKeypointType type) {
	if (keypoints[type.ordinal()].type == type)
		return keypoints[type.ordinal()];

	for (final FacialKeypoint part : keypoints)
		if (part.type == type)
			return part;

	return null;
}
{code}

It seems like there should be some array size check, but this would obviously mean that getting point by {{FacialKeypointType}} is not very reliable since you can get point for other facial feature (e.g. for left eye instead of right one).

I made small change in the code and found that current implementation does not support {{*_CENTER}} features from {{FacialKeypointType}}. It would be good to calculate missing features on the base of the remaining ones (e.g. find centre of the eye by dissecting segment between left and right points).

**Comment by jonhare** [comment]
Original reporter: @sarxos

**Comment by jonhare** [comment]
jonhare:  Hi Bartosz,

Merry Christmas!

I've just committed a fix - I've made KEDetectedFace.getKeypoint(type) delegate directly to FacialKeypoint.getKeypoint(keypoints, type), which does does the looping over the array, but not the the bit that got elements by ordinal (that seemed a little too fragile!).

I've also added a KEDetectedFace.getKeypointInterpolated(FacialKeypointType type) method which works like getKeypoint(type), but will interpolate the positions of the CENTRE points if they are not stored internally.

Let me know if you have any problems.
